File: app/src/main/java/com/example/amapusage/LocationAdapter.java
package com.example.amapusage;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

public class LocationAdapter extends RecyclerView.Adapter<LocationAdapter.DataViewHolder> {
    private Context mContext;
    private ArrayList<String> mList;

    public LocationAdapter() {
    }

    public LocationAdapter(Context mContext, ArrayList<String> mList) {
        this.mContext = mContext;
        this.mList = mList;
    }

    //用于创建ViewHolder
    @Override
    public DataViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(mContext).inflate(R.layout.item, null);
        //使用代码设置宽高（xml布局设置无效时）
        view.setLayoutParams(new ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT));
        DataViewHolder holder = new DataViewHolder(view);
        return holder;
    }

    //绑定数据
    @Override
    public void onBindViewHolder(DataViewHolder holder, int position) {
        holder.tv_data.setText(mList.get(position));
    }

    //数据总数
    @Override
    public int getItemCount() {
        return mList.size();
    }

    //创建ViewHolder
    public static class DataViewHolder extends RecyclerView.ViewHolder {
        TextView tv_data;

        public DataViewHolder(View itemView) {
            super(itemView);
            tv_data = (TextView) itemView.findViewById(R.id.tv_recycle);
            tv_data.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Toast.makeText(App.Companion.getAppContext(), "-----", Toast.LENGTH_SHORT).show();
                }
            });
        }
    }
}

File: app/src/main/java/com/example/amapusage/LocationAdapter.kt
package com.example.amapusage

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CompoundButton
import android.widget.RadioButton
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.item_location.view.*
import java.util.*

class LocationAdapter(private var mContext: Context?, private var mList: ArrayList<String>?) :
    RecyclerView.Adapter<LocationAdapter.DataViewHolder>(),CompoundButton.OnCheckedChangeListener {


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DataViewHolder {
        val view = LayoutInflater.from(mContext).inflate(R.layout.item_location, null)
        return DataViewHolder(view)
    }

    override fun getItemCount(): Int {
        return mList!!.size
    }

    override fun onBindViewHolder(holder: DataViewHolder, position: Int) {
        holder.itemView.locationChecker.tag = position // 标记
        currentHolder = holder
        holder.itemView.tvLocationName.text = mList!![position]
        holder.itemView.locationChecker.setOnCheckedChangeListener(this)
    }

    //创建ViewHolder
    class DataViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView){
        init {
            itemView.setOnClickListener {
                itemView.locationChecker.isChecked = !itemView.locationChecker.isChecked }
        }
    }

    private var currentHolder: DataViewHolder? = null
    var currentCheckPosition :Int = -1
    override fun onCheckedChanged(buttonView: CompoundButton?, isChecked: Boolean) {
        if (currentCheckPosition != -1 && currentHolder?.layoutPosition != currentCheckPosition){
            currentHolder?.itemView?.locationChecker?.isChecked = !isChecked
        }
    }

}
